File: app/src/main/java/com/adsamcik/signalcollector/utility/Network.java
package com.adsamcik.signalcollector.utility;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;

import com.adsamcik.signalcollector.enums.CloudStatus;
import com.franmontiel.persistentcookiejar.ClearableCookieJar;
import com.franmontiel.persistentcookiejar.PersistentCookieJar;
import com.franmontiel.persistentcookiejar.cache.SetCookieCache;
import com.franmontiel.persistentcookiejar.persistence.SharedPrefsCookiePersistor;
import com.google.firebase.crash.FirebaseCrash;

import java.io.IOException;
import java.net.CookieHandler;
import java.net.CookieManager;
import java.net.CookiePolicy;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.CipherSuite;
import okhttp3.ConnectionSpec;
import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.Headers;
import okhttp3.HttpUrl;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.TlsVersion;

public final class Network {
	public static final String URL_DATA_UPLOAD = Server.URL_DATA_UPLOAD;
	public static final String URL_TILES = Server.URL_TILES;
	public static final String URL_USER_STATS = Server.URL_USER_STATS;
	public static final String URL_STATS = Server.URL_STATS;
	public static final String URL_MAPS_AVAILABLE = Server.URL_MAPS_AVAILABLE;
	public static final String URL_FEEDBACK = Server.URL_FEEDBACK;

	public static CloudStatus cloudStatus;

	private static ConnectionSpec getSpec() {
		return new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
				.tlsVersions(TlsVersion.TLS_1_2)
				.cipherSuites(CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
						CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
						CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
						CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA)
				.build();
	}

	private static CookieJar getCookieJar(@NonNull Context context) {
		return new PersistentCookieJar(new SetCookieCache(), new SharedPrefsCookiePersistor(context));
	}

	public static OkHttpClient client(final String userToken, final Context context) {
		return new OkHttpClient.Builder()
				.connectionSpecs(Collections.singletonList(getSpec()))
				.cookieJar(getCookieJar(context))
				.authenticator((route, response) ->
						response.request().newBuilder()
								.headers(new Headers.Builder().add("userToken", userToken).add("manufacturer", Build.MANUFACTURER).add("model", Build.MODEL).build())
								.build())
				.build();
	}

	public static OkHttpClient client(final Context context) {
		return new OkHttpClient.Builder()
				.connectionSpecs(Collections.singletonList(getSpec()))
				.cookieJar(getCookieJar(context))
				.build();
	}

	public static Request request(final String url, final RequestBody body) {
		return new Request.Builder().url(url).post(body).build();
	}

	public static MultipartBody.Builder generateAuthBody(@NonNull final String userToken) {
		return new MultipartBody.Builder()
				.setType(MultipartBody.FORM)
				.addFormDataPart("userToken", userToken)
				.addFormDataPart("manufacturer", Build.MANUFACTURER)
				.addFormDataPart("model", Build.MODEL);
	}

	public static void register(final String userToken, final String token, @NonNull final Context context) {
		if (userToken != null && token != null && !Assist.isEmulator())
			register(userToken, "token", token, Preferences.PREF_SENT_TOKEN_TO_SERVER, Server.URL_TOKEN_REGISTRATION, context);
	}

	private static void register(@NonNull final String userToken, @NonNull final String valueName, @NonNull final String value, @NonNull final String preferencesName, @NonNull final String url, @NonNull Context context) {
		RequestBody formBody = generateAuthBody(userToken)
				.addFormDataPart(valueName, value)
				.build();
		Request request = request(url, formBody);
		client(userToken, context).newCall(request).enqueue(new Callback() {
			@Override
			public void onFailure(@NonNull Call call, @NonNull IOException e) {
				FirebaseCrash.log("Register " + preferencesName);
				FirebaseCrash.report(e);
			}

			@Override
			public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
				Preferences.get().edit().putBoolean(preferencesName, true).apply();
				response.close();
			}
		});
	}

	public static String generateVerificationString(String uid, Long length) {
		return Server.generateVerificationString(uid, length);
	}
}

File: network/src/main/java/com/adsamcik/signals/network/Network.kt
package com.adsamcik.signals.network

import android.content.Context
import android.os.Build
import com.adsamcik.signals.utilities.Preferences
import com.adsamcik.signals.utilities.test.useMock
import com.crashlytics.android.Crashlytics
import com.franmontiel.persistentcookiejar.PersistentCookieJar
import com.franmontiel.persistentcookiejar.cache.SetCookieCache
import com.franmontiel.persistentcookiejar.persistence.SharedPrefsCookiePersistor
import okhttp3.*
import java.io.IOException
import java.util.*

object Network {
    private const val TAG = "SignalsNetwork"
    val URL_DATA_UPLOAD = Server.URL_DATA_UPLOAD
    val URL_TILES = Server.URL_TILES
    val URL_PERSONAL_TILES = Server.URL_PERSONAL_TILES
    val URL_USER_STATS = Server.URL_USER_STATS
    val URL_STATS = Server.URL_STATS
    val URL_GENERAL_STATS = Server.URL_GENERAL_STATS
    val URL_MAPS_AVAILABLE = Server.URL_MAPS_AVAILABLE
    val URL_FEEDBACK = Server.URL_FEEDBACK
    val URL_USER_INFO = Server.URL_USER_INFO
    val URL_USER_PRICES = Server.URL_USER_PRICES
    val URL_CHALLENGES_LIST = Server.URL_CHALLENGES_LIST

    val URL_USER_UPDATE_MAP_PREFERENCE = Server.URL_USER_UPDATE_MAP_PREFERENCE
    val URL_USER_UPDATE_PERSONAL_MAP_PREFERENCE = Server.URL_USER_UPDATE_PERSONAL_MAP_PREFERENCE

    @CloudStatus
    var cloudStatus = CloudStatus.UNKNOWN

    private var cookieJar: PersistentCookieJar? = null

    private val spec: ConnectionSpec
        get() = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
                .tlsVersions(TlsVersion.TLS_1_2)
                .cipherSuites(CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
                        CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
                        CipherSuite.TLS_DHE_RSA_WITH_AES_256_CBC_SHA)
                .build()

    private fun getCookieJar(context: Context): CookieJar {
        if (cookieJar == null)
            cookieJar = PersistentCookieJar(SetCookieCache(), SharedPrefsCookiePersistor(context))
        return cookieJar!!
    }

    fun clearCookieJar(context: Context) {
        if (cookieJar == null)
            getCookieJar(context)
        cookieJar!!.clear()
    }

    fun client(context: Context, userToken: String?): OkHttpClient {
        return if (userToken == null) client(context)
        else
            OkHttpClient.Builder()
                    .connectionSpecs(Collections.singletonList(spec))
                    .cookieJar(getCookieJar(context))
                    .authenticator({ _, response ->
                        if (response.request().header("userToken") != null)
                            return@authenticator null
                        else {
                            response.request().newBuilder().header("userToken", userToken).build()
                        }
                    }).build()
    }

    private fun client(context: Context): OkHttpClient {
        return OkHttpClient.Builder()
                .connectionSpecs(listOf(spec))
                .cookieJar(getCookieJar(context))
                .build()
    }

    fun requestGET(url: String): Request = Request.Builder().url(url).build()

    fun requestPOST(url: String, body: RequestBody): Request =
            Request.Builder().url(url).post(body).build()

    fun generateAuthBody(userToken: String): MultipartBody.Builder {
        return MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("userToken", userToken)
                .addFormDataPart("manufacturer", Build.MANUFACTURER)
                .addFormDataPart("model", Build.MODEL)
    }

    fun register(context: Context, userToken: String, token: String) {
        if (!useMock)
            register(context, userToken, "token", token, Preferences.PREF_SENT_TOKEN_TO_SERVER, Server.URL_TOKEN_REGISTRATION)
    }

    private fun register(context: Context, userToken: String, valueName: String, value: String, preferencesName: String, url: String) {
        val formBody = generateAuthBody(userToken)
                .addFormDataPart(valueName, value)
                .build()
        val request = requestPOST(url, formBody)
        client(context, userToken).newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Crashlytics.log("Register " + preferencesName)
                Crashlytics.logException(e)
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                Preferences.getPref(context).edit().putBoolean(preferencesName, true).apply()
                response.close()
            }
        })
    }

    fun generateVerificationString(uid: String, length: Long?): String =
            Server.generateVerificationString(uid, length)
}